Analysis/DataStorageProjects/Delfin/CentralizedManager/CentralizedManager.md

```diff
@@ -0,0 +1,166 @@
+
+  Invalid id for CM shall report error
+
+- Users of delfin should be able to update CM Access Info details for a registered CM, providing its id
+
+  Update shall verify the Access Info and if verification is success, update the CM details for the id of the CM, that is currently managed by delfin
+
+  Invalid Access Info for CM shall report error and no changes shall be done to delfin
+
+- Users of delfin should be able to SYNC all registered CMs for updating list of managed storage arrays of each CM
+
+  When user calls SYNC all, delfin with re-discover all the storage arrays from CM and update managed storage arrays
+
+- Users of delfin should be able to SYNC a specific CM with the its id, for updating list of managed storage arrays by that CM
+
+  When user calls SYNC for a CM, delfin with re-discover all the storage arrays from the CM and update managed storage arrays
+
+### Non Functional Requirements
+
+- Centralized Manager feature needs to be generic to support different storage vendors
+- Centralized Manager feature needs to be compatible with already existing APIs of Delfin
+
+## Architecture Analysis
+
+Adding support for CM requires the addition of CM related APIs to current delfin APIs. Driver Manager and Driver API will require update. Impacted modules of delfin are highlighted below
+
+![Centralized Managers Architecture](Delfin.png)
+
+### High Level Design
+
+Add the Centralized Manager as a separate entity with APIs to support Create, Delete and Get operations. Access Info for CM can be with different Schema and supports APIs Update/Get. APIs for CM sync and sync-all also will be added for updating managed storages by CMs.
+
+- Add REST APIs to support CRUD operations of CM and its access info.
+- Update framework, driver manager APIs for supporting CM drivers
+- None of the existing REST APIs will be impacted, existing delfin users will not require changes for existing features
+
+#### Delfin DB Schema changes for CM
+
+Changes in Delfin DB Schema to support the CM is captured here.
+
+These changes will require migrating or re-creating of the DB for existing users of Delfin.
+
+##### Centralized Manager model
+
+A new CM Model will be added to the Delfin schemas
+
+Attributes | Type | Description/enum
+-- | -- | --
+id | string | UUID of the Centralized Manager
+vendor | string | Vendor of the CM
+model | string | Model of the CM
+resources | JSON | Array of storages the associated CM
+
+#### New APIs for CM
+
+Following new REST APIs will be added to Delfin REST APIs to support CM
+
+```sh
+POST /centralized-managers -d <cm-access-info>
+
+DELETE /centralized-managers/<cm-id>
+
+GET /centralized-managers
+
+GET /centralized-managers/<cm-id>
+
+PUT /centralized-managers/access-info -d <cm-access-info>
+
+GET /centralized-managers/access-info
+
+PUT /centralized-managers/sync-all
+
+PUT /centralized-managers/<cm-id>/sync
+```
+
+#### Sequence Diagrams
+
+- **Register Centralized Manager**
```

Comment:
Some suggestions to seq diagram
- rename :CMctrl to API/CMctrlr
 - Request goes  goes from driver to back end during CM registration is Get_LIST_OF_STORAGES 
 - loop to iterate through storage list and register each storage individually should be in API/CMctrlr
 - individual storage registration happens inside the loop is not captured in the sequence, we can either show it full, or add a ref here . same applies to sync loop
 -

Reply:
> Some suggestions to seq diagram
> 
> * rename :CMctrl to API/CMctrlr
Ok

> * Request goes  goes from driver to back end during CM registration is Get_LIST_OF_STORAGES
This depends on driver implementation and we do not specify this interface. In future, user may supply list of storages to register by CM

> * loop to iterate through storage list and register each storage individually should be in API/CMctrlr
This is handled in discover_cm() function in, drivers/api.py similar to how storage do discover_storage(). 

> * individual storage registration happens inside the loop is not captured in the sequence, we can either show it full, or add a ref here . same applies to sync loop
Ok